web/client-api/src/main/java/io/deephaven/web/client/api/barrage/def/TableAttributesDefinition.java

```diff
@@ -0,0 +1,81 @@
+package io.deephaven.web.client.api.barrage.def;
+
+import io.deephaven.web.shared.data.RollupDefinition;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Set;
+
+public class TableAttributesDefinition {
+    public static final String INPUT_TABLE_ATTRIBUTE = "InputTable",
+            TOTALS_TABLE_ATTRIBUTE = "TotalsTable",
+            TABLE_DESCRIPTION_ATTRIBUTE = "TableDescription",
+            HIERARCHICAL_SOURCE_TABLE_ATTRIBUTE = "HierarchicalSourceTable",
+            HIERARCHICAL_SOURCE_INFO_ATTRIBUTE = "HierarchicalSourceTableInfo",
+            LAYOUT_HINTS_ATTRIBUTE = "LayoutHints",
+            PLUGIN_NAME = "PluginName";
+
+    private static final String HIERARCHICAL_COLUMN_NAME = "hierarchicalColumnName";
+
+    // special cased attributes that have a complex type yet are always sent
+    private RollupDefinition rollupDefinition;// rollup subtype of "HierarchicalSourceTableInfo"
+
+    private final Map<String, String> map;
+    private final Set<String> remainingAttributeKeys;
+
+    public TableAttributesDefinition(Map<String, String> keys, Set<String> remainingAttributes) {
+        map = keys;
+        this.remainingAttributeKeys = remainingAttributes;
+        if (map.containsKey(HIERARCHICAL_SOURCE_INFO_ATTRIBUTE + "." + HIERARCHICAL_COLUMN_NAME)) {
+            // marker present for tree table metadata
+            rollupDefinition = new RollupDefinition();
+            rollupDefinition.setByColumns(map.get(HIERARCHICAL_COLUMN_NAME + "." + "byColumns").split(","));
+            rollupDefinition.setLeafType(
+                    RollupDefinition.LeafType.valueOf(map.get(HIERARCHICAL_COLUMN_NAME + "." + "leafType")));
+        }
```

Comment:
I'm a little confused about the constants being used here... `HIERARCHICAL_COLUMN_NAME` is `drinkingCamelCase`, and here on the first line you get `HIERARCHICAL_SOURCE_INFO_ATTRIBUTE + "." + HIERARCHICAL_COLUMN_NAME`, or getting `"HierarchicalSourceTableInfo.hierarchicalColumnName"`, which makes sense... but then below, you're getting `"hierarchicalColumnName.byColumns"` - does that not need the `HIERACRCHICAL_SOURCE_INFO_ATTRIBUTE` prefix? If not why not?

I think you should just define these all as constants so it's not as confusing:
```
SOURCE_INFO_COLUMN_NAME = HIERARCHICAL_SOURCE_INFO_ATTRIBUTE + "." + HIERARCHICAL_COLUMN_NAME;
BY_COLUMN_ATTRIBUTE = HIERARCHICAL_COLUMN_NAME + ".byColumns";
LEAF_TYPE_ATTRIBUTE = HIERARCHICAL_COLUMN_NAME + ".leafType"

```

Reply:
Yeah the real headache is that they get defined twice, once on the server and once on the client. As discussed in the custom object call, let's redo this again once that is merged? I will give this a quick look though to make sure you're right, but trees don't work anyway...